File: agents/go-agent/agent.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"runtime"
	"sync"
	"sync/atomic"
	"time"
)

// EventMetadata contains metadata for all events
type EventMetadata struct {
	TraceID        string `json:"trace_id"`
	ProcessID      int    `json:"process_id"`
	ThreadID       uint64 `json:"thread_id"`
	TimestampNs    int64  `json:"timestamp_ns"`
	SequenceNumber uint64 `json:"sequence_number"`
}

// GoroutineCreateEvent represents goroutine creation
type GoroutineCreateEvent struct {
	Metadata         EventMetadata `json:"metadata"`
	NewGoroutineID   uint64        `json:"new_goroutine_id"`
	ParentGoroutineID uint64       `json:"parent_goroutine_id"`
	EventType        string        `json:"event_type"`
}

// FunctionCallEvent represents a function call
type FunctionCallEvent struct {
	Metadata     EventMetadata `json:"metadata"`
	FunctionName string        `json:"function_name"`
	Module       string        `json:"module"`
	StackDepth   int           `json:"stack_depth"`
	EventType    string        `json:"event_type"`
}

// ChannelOpEvent represents a channel operation
type ChannelOpEvent struct {
	Metadata  EventMetadata `json:"metadata"`
	ChannelID uint64        `json:"channel_id"`
	Operation string        `json:"operation"` // "send" or "receive"
	EventType string        `json:"event_type"`
}

// Agent is the Go tracing agent
type Agent struct {
	traceID        string
	serverAddr     string
	conn           net.Conn
	sequenceNumber uint64
	enabled        bool
	mu             sync.Mutex
}

// NewAgent creates a new Go agent
func NewAgent(traceID, serverAddr string) *Agent {
	return &Agent{
		traceID:    traceID,
		serverAddr: serverAddr,
		enabled:    false,
	}
}

// Connect establishes connection to the trace server
func (a *Agent) Connect() error {
	conn, err := net.Dial("tcp", a.serverAddr)
	if err != nil {
		return fmt.Errorf("failed to connect to trace server: %w", err)
	}
	a.conn = conn
	fmt.Printf("[Aevum] Connected to trace server at %s\n", a.serverAddr)
	return nil
}

// Disconnect closes the connection
func (a *Agent) Disconnect() {
	if a.conn != nil {
		a.conn.Close()
		a.conn = nil
	}
}

// SendEvent sends an event to the trace server
func (a *Agent) SendEvent(event interface{}) error {
	if a.conn == nil {
		return fmt.Errorf("not connected to trace server")
	}

	data, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("failed to marshal event: %w", err)
	}

	// Send length prefix
	length := uint32(len(data))
	if err := binary.Write(a.conn, binary.BigEndian, length); err != nil {
		return fmt.Errorf("failed to write length: %w", err)
	}

	// Send data
	if _, err := a.conn.Write(data); err != nil {
		return fmt.Errorf("failed to write data: %w", err)
	}

	return nil
}

// CreateMetadata creates event metadata
func (a *Agent) CreateMetadata() EventMetadata {
	seq := atomic.AddUint64(&a.sequenceNumber, 1)
	return EventMetadata{
		TraceID:        a.traceID,
		ProcessID:      os.Getpid(),
		ThreadID:       getGoroutineID(),
		TimestampNs:    time.Now().UnixNano(),
		SequenceNumber: seq,
	}
}

// RecordFunctionCall records a function call
func (a *Agent) RecordFunctionCall(functionName, module string, stackDepth int) {
	if !a.enabled {
		return
	}

	event := FunctionCallEvent{
		Metadata:     a.CreateMetadata(),
		FunctionName: functionName,
		Module:       module,
		StackDepth:   stackDepth,
		EventType:    "FunctionCall",
	}

	if err := a.SendEvent(event); err != nil {
		fmt.Printf("[Aevum] Failed to send event: %v\n", err)
	}
}

// RecordGoroutineCreate records goroutine creation
func (a *Agent) RecordGoroutineCreate(newGoroutineID, parentGoroutineID uint64) {
	if !a.enabled {
		return
	}

	event := GoroutineCreateEvent{
		Metadata:          a.CreateMetadata(),
		NewGoroutineID:    newGoroutineID,
		ParentGoroutineID: parentGoroutineID,
		EventType:         "GoroutineCreate",
	}

	if err := a.SendEvent(event); err != nil {
		fmt.Printf("[Aevum] Failed to send event: %v\n", err)
	}
}

// RecordChannelOp records a channel operation
func (a *Agent) RecordChannelOp(channelID uint64, operation string) {
	if !a.enabled {
		return
	}

	event := ChannelOpEvent{
		Metadata:  a.CreateMetadata(),
		ChannelID: channelID,
		Operation: operation,
		EventType: "ChannelOp",
	}

	if err := a.SendEvent(event); err != nil {
		fmt.Printf("[Aevum] Failed to send event: %v\n", err)
	}
}

// Start starts the agent
func (a *Agent) Start() error {
	if err := a.Connect(); err != nil {
		return err
	}
	a.enabled = true
	fmt.Printf("[Aevum] Go agent started (trace_id: %s)\n", a.traceID)
	return nil
}

// Stop stops the agent
func (a *Agent) Stop() {
	a.enabled = false
	a.Disconnect()
	fmt.Println("[Aevum] Go agent stopped")
}

// getGoroutineID returns the current goroutine ID
func getGoroutineID() uint64 {
	// This is a simplified implementation
	// In production, you'd use runtime.Stack() parsing or unsafe tricks
	return uint64(runtime.NumGoroutine())
}

// Global agent instance
var globalAgent *Agent
var agentMu sync.Mutex

// Attach attaches the agent to the current process
func Attach(traceID, serverAddr string) (*Agent, error) {
	agentMu.Lock()
	defer agentMu.Unlock()

	if globalAgent != nil {
		return globalAgent, fmt.Errorf("agent already attached")
	}

	agent := NewAgent(traceID, serverAddr)
	if err := agent.Start(); err != nil {
		return nil, err
	}

	globalAgent = agent
	return agent, nil
}

// Detach detaches the agent
func Detach() {
	agentMu.Lock()
	defer agentMu.Unlock()

	if globalAgent != nil {
		globalAgent.Stop()
		globalAgent = nil
	}
}

// GetAgent returns the global agent instance
func GetAgent() *Agent {
	agentMu.Lock()
	defer agentMu.Unlock()
	return globalAgent
}

File: agents/node-agent/agent.js
const inspector = require('inspector');
const { EventEmitter } = require('events');
const net = require('net');

/**
 * Aevum Node.js Agent - Inspector Protocol integration
 * 
 * Uses the V8 Inspector Protocol to capture execution events
 */

class EventMetadata {
    constructor(traceId, sequenceNumber) {
        this.trace_id = traceId;
        this.process_id = process.pid;
        this.thread_id = 0; // Node.js is single-threaded (main thread)
        this.timestamp_ns = process.hrtime.bigint().toString();
        this.sequence_number = sequenceNumber;
    }
}

class AevumNodeAgent extends EventEmitter {
    constructor(traceId, serverHost = 'localhost', serverPort = 9876) {
        super();
        this.traceId = traceId;
        this.serverHost = serverHost;
        this.serverPort = serverPort;
        this.sequenceNumber = 0;
        this.enabled = false;
        this.socket = null;
        this.session = null;
    }

    /**
     * Connect to the trace server
     */
    connect() {
        return new Promise((resolve, reject) => {
            this.socket = net.createConnection({
                host: this.serverHost,
                port: this.serverPort
            }, () => {
                console.log(`[Aevum] Connected to trace server at ${this.serverHost}:${this.serverPort}`);
                resolve();
            });

            this.socket.on('error', (err) => {
                console.error(`[Aevum] Socket error: ${err.message}`);
                reject(err);
            });
        });
    }

    /**
     * Disconnect from the trace server
     */
    disconnect() {
        if (this.socket) {
            this.socket.end();
            this.socket = null;
        }
    }

    /**
     * Send an event to the trace server
     */
    sendEvent(event) {
        if (!this.socket) {
            return;
        }

        try {
            const eventJson = JSON.stringify(event);
            const eventBuffer = Buffer.from(eventJson, 'utf-8');

            // Send length prefix (4 bytes, big-endian) + data
            const lengthBuffer = Buffer.allocUnsafe(4);
            lengthBuffer.writeUInt32BE(eventBuffer.length, 0);

            this.socket.write(Buffer.concat([lengthBuffer, eventBuffer]));
        } catch (err) {
            console.error(`[Aevum] Failed to send event: ${err.message}`);
        }
    }

    /**
     * Create event metadata
     */
    createMetadata() {
        this.sequenceNumber++;
        return new EventMetadata(this.traceId, this.sequenceNumber);
    }

    /**
     * Handle debugger events
     */
    handleDebuggerEvent(method, params) {
        if (!this.enabled) {
            return;
        }

        try {
            if (method === 'Debugger.paused') {
                this.handleFunctionCall(params);
            } else if (method === 'Runtime.executionContextCreated') {
                this.handleExecutionContext(params);
            } else if (method === 'Runtime.consoleAPICalled') {
                this.handleConsoleAPI(params);
            }
        } catch (err) {
            console.error(`[Aevum] Error handling debugger event: ${err.message}`);
        }
    }

    /**
     * Handle function call event
     */
    handleFunctionCall(params) {
        const callFrame = params.callFrames && params.callFrames[0];
        if (!callFrame) return;

        const event = {
            event_type: 'FunctionCall',
            metadata: this.createMetadata(),
            function_name: callFrame.functionName || '<anonymous>',
            module: callFrame.url || '<unknown>',
            line: callFrame.location.lineNumber,
            column: callFrame.location.columnNumber,
            stack_depth: params.callFrames.length
        };

        this.sendEvent(event);
    }

    /**
     * Handle execution context creation
     */
    handleExecutionContext(params) {
        const event = {
            event_type: 'ExecutionContextCreated',
            metadata: this.createMetadata(),
            context_id: params.context.id,
            context_name: params.context.name,
            origin: params.context.origin
        };

        this.sendEvent(event);
    }

    /**
     * Handle console API calls
     */
    handleConsoleAPI(params) {
        const event = {
            event_type: 'ConsoleAPI',
            metadata: this.createMetadata(),
            type: params.type,
            args: params.args.map(arg => arg.value || arg.description || '<object>')
        };

        this.sendEvent(event);
    }

    /**
     * Start the agent
     */
    async start() {
        await this.connect();

        // Open inspector session
        this.session = new inspector.Session();
        this.session.connect();

        // Enable debugging domains
        this.session.post('Debugger.enable');
        this.session.post('Runtime.enable');
        this.session.post('Profiler.enable');

        // Set up event listeners
        this.session.on('Debugger.paused', (params) => {
            this.handleDebuggerEvent('Debugger.paused', params);
            // Resume execution
            this.session.post('Debugger.resume');
        });

        this.session.on('Runtime.executionContextCreated', (params) => {
            this.handleDebuggerEvent('Runtime.executionContextCreated', params);
        });

        this.session.on('Runtime.consoleAPICalled', (params) => {
            this.handleDebuggerEvent('Runtime.consoleAPICalled', params);
        });

        this.enabled = true;
        console.log(`[Aevum] Node.js agent started (trace_id: ${this.traceId})`);
    }

    /**
     * Stop the agent
     */
    stop() {
        this.enabled = false;

        if (this.session) {
            this.session.post('Debugger.disable');
            this.session.post('Runtime.disable');
            this.session.post('Profiler.disable');
            this.session.disconnect();
            this.session = null;
        }

        this.disconnect();
        console.log('[Aevum] Node.js agent stopped');
    }
}

// Global agent instance
let globalAgent = null;

/**
 * Attach the Aevum agent to the current process
 */
async function attach(traceId, serverHost = 'localhost', serverPort = 9876) {
    if (globalAgent) {
        console.log('[Aevum] Agent already attached');
        return globalAgent;
    }

    globalAgent = new AevumNodeAgent(traceId, serverHost, serverPort);
    await globalAgent.start();
    return globalAgent;
}

/**
 * Detach the Aevum agent
 */
function detach() {
    if (globalAgent) {
        globalAgent.stop();
        globalAgent = null;
    }
}

module.exports = {
    AevumNodeAgent,
    attach,
    detach
};
